js/src/common/components/Tooltip.tsx

```diff
@@ -0,0 +1,125 @@
+import Component, { ComponentAttrs } from '../Component';
+import type Mithril from 'mithril';
+import classList from '../utils/classList';
+import { TooltipOptions } from '../../../@types/tooltips';
+
+export interface TooltipAttrs extends ComponentAttrs {
+  /**
+   * Tooltip textual content.
+   */
+  text: string;
+  /**
+   * If inline, uses a `<span>` container, else uses a `<div>`.
+   *
+   * Default: `false`.
+   */
+  inline?: boolean;
+  /**
+   * Manually show tooltip. `false` will show based on cursor events.
+   *
+   * Default: `false`.
+   */
+  tooltipVisible?: boolean;
+  /**
+   * Whether to show on focus.
+   *
+   * Default: `true`.
+   */
+  showOnFocus?: boolean;
+  /**
+   * Tooltip position around element.
+   */
+  position?: 'top' | 'bottom' | 'left' | 'right';
+  /**
+   * Whether HTML content is allowed in the tooltip.
+   *
+   * **Warning:** this is a possible XSS attack vector. This option shouldn't
+   * be used wherever possible, and will not work when we migrate to CSS-only
+   * tooltips.
+   *
+   * @deprecated
+   */
+  html?: boolean;
+  /**
+   * Sets the delay between a trigger state occurring and the tooltip appearing
+   * on-screen.
+   *
+   * **Warning:** this option will be removed when we switch to CSS-only
+   * tooltips.
+   *
+   * @deprecated
+   */
+  delay?: number;
+}
+
+/**
+ * The `Tooltip` component is used to create a tooltip for an element. It
+ * surrounds it with a div (or span) which has the required tooltip setup
+ * applied.
+ *
+ * You should use this for any tooltips you create to allow for backwards
+ * compatibility when we switch to pure CSS tooltips instead of Bootstrap
+ * tooltips.
+ *
+ * @example <caption>Basic usage</caption>
+ *          <Tooltip text="You wish!">
+ *            <Button>
+ *              Click for free money!
+ *            </Button>
+ *          </Tooltip>
+ *
+ * @example <caption>Use of `position` and `showOnFocus` attrs</caption>
+ *          <Tooltip text="Woah! That's cool!" position="bottom" showOnFocus>
+ *            <div>3 replies</div>
+ *          </Tooltip>
+ */
+export default class Tooltip extends Component<TooltipAttrs> {
+  view(vnode) {
+    const { children } = vnode;
+
+    // We remove these to get the remaining attrs to pass to the DOM element
+    const { text, inline, tooltipVisible, showOnFocus, position, html, delay, ...attrs } = this.attrs;
+
+    if (inline) {
```

Comment:
Could we return the children directly instead of wrapping it in an additional level of div/span?

Reply:
What challenges could that pose? I'm not familiar with this particular technology

Reply:
If we have:
```html
<div data-tooltip aria-label="Very cool">
  Cool beans!
</div>
```

Where that `<div>` already has `::before` or `::after` styles, that would interfere with with the pseudoelement styles defined by the use of `data-tooltip` on it. This could result in very odd tooltip styling due to conflicts.

I initially made this component to work around this issue by wrapping the element in a container.

Maybe a `wrap: 'block' | 'inline'` attr would be better?

Reply:
Since you've been working a lot more with this than I have, is there a danger to inserting unexpected wrapping elements other than breakage of `>` css specifiers?

Reply:
Not that I can think of off of the top of my head, no.

Reply:
> Maybe we could allow for the ability to specify a custom container element, then create that using m()?

This seems overly hacky IMO. Let's try the `div` and `span` change.

Reply:
Just had to swap it out for using `<div>` with `inline-block` due to an issue with padding on `display: inline`.

`inline-block` seems to work in all areas of core so far.